crates/server/src/rpc/mod.rs

```diff
@@ -121,3 +146,60 @@ async fn handle_rejection(err: Rejection) -> Result<impl Reply, std::convert::In
         ))
     }
 }
+
+pub enum AuthenticationErrors {
+    FailedToSendChallenge,
+    WrongSignature,
+    Timeout,
+    NotTextMessage,
+    UnexpectedError(Box<dyn std::error::Error + Send + Sync>),
+}
+
+pub async fn authenticate_dcl_user(
+    ws: WebSocket,
+) -> Result<(WebSocket, Address), AuthenticationErrors> {
+    let authenticator = Authenticator::new();
+    let (mut ws_write, mut ws_read) = ws.split();
+
+    let message_to_be_firmed = format!("signature_challenge_{}", fastrand::u32(..));
+
+    if ws_write
+        .send(Message::text(&message_to_be_firmed))
+        .await
+        .is_err()
+    {
+        return Err(AuthenticationErrors::FailedToSendChallenge);
+    }
+
+    match tokio::time::timeout(Duration::from_secs(30), ws_read.next()).await {
+        Ok(client_response) => {
+            let response = client_response.unwrap().unwrap();
+            if let Ok(auth_chain) = response.to_str() {
+                let auth_chain = AuthChain::from_json(auth_chain).unwrap();
+                if let Ok(address) = authenticator
+                    .verify_signature(&auth_chain, &message_to_be_firmed)
+                    .await
+                {
+                    let address = address.to_owned();
+                    match ws_write.reunite(ws_read) {
+                        Ok(ws) => Ok((ws, address)),
+                        Err(err) => Err(AuthenticationErrors::UnexpectedError(Box::new(err))),
+                    }
+                } else if let Err(err) = ws_write.close().await {
```

Comment:
What do you say if we close the websocket in the upgrade handler and return the error first?

Reply:
Okay but the error would be `Err((Websocket,AuthenticationErrors))` because we have to consume the ws